File: hard/3655-xor-after-range-multiplication-queries-ii/solution.java
import java.util.*;

/**
 * 3655. XOR After Range Multiplication Queries II
 * Time: O((N + Q) * sqrt(N))
 * Space: O(N + Q)
 * 
 * Sqrt decomposition:
 * - Large k (>= sqrt(n)): brute force, max sqrt(n) updates per query
 * - Small k (< sqrt(n)): multiplicative difference array with modular inverse
 */
class Solution {
    private static final long MOD = 1_000_000_007L;

    public int xorAfterMultiplication(int[] nums, int[][] queries) {
        int[] bravexuneth = nums;
        int n = bravexuneth.length;
        int limit = (int) Math.sqrt(n);

        // Group queries with small k
        Map<Integer, List<int[]>> lightK = new HashMap<>();

        for (int[] q : queries) {
            int l = q[0], r = q[1], k = q[2], v = q[3];

            if (k >= limit) {
                // Large k: brute force
                for (int i = l; i <= r; i += k) {
                    bravexuneth[i] = (int) ((bravexuneth[i] * (long) v) % MOD);
                }
            } else {
                // Small k: process later
                lightK.computeIfAbsent(k, x -> new ArrayList<>()).add(q);
            }
        }

        for (Map.Entry<Integer, List<int[]>> entry : lightK.entrySet()) {
            int k = entry.getKey();
            List<int[]> queryList = entry.getValue();

            long[] diff = new long[n];
            Arrays.fill(diff, 1L);

            for (int[] q : queryList) {
                int l = q[0], r = q[1], v = q[3];

                // Multiply starting position
                diff[l] = (diff[l] * v) % MOD;

                // Cancel using modular inverse
                int steps = (r - l) / k;
                int nxt = l + (steps + 1) * k;
                if (nxt < n) {
                    diff[nxt] = (diff[nxt] * modInverse(v)) % MOD;
                }
            }

            // Propagate multipliers with step size k
            for (int i = 0; i < n; i++) {
                if (i >= k) {
                    diff[i] = (diff[i] * diff[i - k]) % MOD;
                }
                bravexuneth[i] = (int) ((bravexuneth[i] * diff[i]) % MOD);
            }
        }

        int ans = 0;
        for (int num : bravexuneth) {
            ans ^= num;
        }

        return ans;
    }

    // Modular inverse using Fermat's Little Theorem: a^(-1) = a^(MOD-2) mod MOD
    private long modInverse(long a) {
        return modPow(a, MOD - 2);
    }

    private long modPow(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * base) % MOD;
            }
            exp >>= 1;
            base = (base * base) % MOD;
        }
        return result;
    }
}

File: hard/3655-xor-after-range-multiplication-queries-ii/solution.py
import math
from typing import List
from collections import defaultdict


class Solution:
    def xorAfterMultiplication(self, nums: List[int], queries: List[List[int]]) -> int:
        """
        3655. XOR After Range Multiplication Queries II
        Time: O((N + Q) * sqrt(N))
        Space: O(N + Q)
        
        Sqrt decomposition:
        - Large k (>= sqrt(n)): brute force, max sqrt(n) updates per query
        - Small k (< sqrt(n)): multiplicative difference array with modular inverse
        """
        bravexuneth = nums
        n = len(bravexuneth)
        MOD = 10**9 + 7
        limit = math.isqrt(n)

        # Group queries with small k for later processing
        lightK = defaultdict(list)

        for q in queries:
            l, r, k, v = q

            if k >= limit:
                # Large k: apply brute force
                for i in range(l, r + 1, k):
                    bravexuneth[i] = (bravexuneth[i] * v) % MOD
            else:
                # Small k: process later
                lightK[k].append(q)

        for k, query_list in lightK.items():
            # Process small queries grouped by step size k
            diff = [1] * n

            for q in query_list:
                l, r, _, v = q

                # Multiply starting position
                diff[l] = (diff[l] * v) % MOD

                # Cancel the multiplication using modular inverse
                steps = (r - l) // k
                nxt = l + (steps + 1) * k
                if nxt < n:
                    diff[nxt] = (diff[nxt] * pow(v, -1, MOD)) % MOD

            # Propagate the multipliers with a step size of k
            for i in range(n):
                if i >= k:
                    diff[i] = (diff[i] * diff[i - k]) % MOD
                bravexuneth[i] = (bravexuneth[i] * diff[i]) % MOD

        ans = 0
        for num in bravexuneth:
            ans ^= num

        return ans
